File: bindings/swift/Sources/UMICP/Discovery/OperationSchema.swift
import Foundation

/// Operation schema compatible with MCP JSON Schema
public struct OperationSchema: Codable, Equatable {
    public let name: String
    public var title: String?
    public var description: String?
    public let inputSchema: [String: AnyCodable]
    public var outputSchema: [String: AnyCodable]?
    public var annotations: [String: AnyCodable]?
    
    enum CodingKeys: String, CodingKey {
        case name, title, description
        case inputSchema = "input_schema"
        case outputSchema = "output_schema"
        case annotations
    }
    
    public init(
        name: String,
        inputSchema: [String: AnyCodable],
        title: String? = nil,
        description: String? = nil,
        outputSchema: [String: AnyCodable]? = nil,
        annotations: [String: AnyCodable]? = nil
    ) {
        self.name = name
        self.inputSchema = inputSchema
        self.title = title
        self.description = description
        self.outputSchema = outputSchema
        self.annotations = annotations
    }
    
    /// Builder pattern for title
    public func withTitle(_ title: String) -> OperationSchema {
        var copy = self
        copy.title = title
        return copy
    }
    
    /// Builder pattern for description
    public func withDescription(_ description: String) -> OperationSchema {
        var copy = self
        copy.description = description
        return copy
    }
    
    /// Builder pattern for output schema
    public func withOutputSchema(_ schema: [String: AnyCodable]) -> OperationSchema {
        var copy = self
        copy.outputSchema = schema
        return copy
    }
    
    /// Builder pattern for annotations
    public func withAnnotations(_ annotations: [String: AnyCodable]) -> OperationSchema {
        var copy = self
        copy.annotations = annotations
        return copy
    }
    
    /// Convert to dictionary representation
    public func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "name": name,
            "input_schema": inputSchema.mapValues { $0.value }
        ]
        
        if let title = title {
            dict["title"] = title
        }
        if let description = description {
            dict["description"] = description
        }
        if let outputSchema = outputSchema {
            dict["output_schema"] = outputSchema.mapValues { $0.value }
        }
        if let annotations = annotations {
            dict["annotations"] = annotations.mapValues { $0.value }
        }
        
        return dict
    }
}

File: bindings/csharp/Umicp.Core/ToolDiscovery/OperationSchema.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Umicp.SDK.ToolDiscovery;

/// <summary>
/// Operation schema compatible with MCP JSON Schema
/// </summary>
public class OperationSchema
{
    /// <summary>
    /// Operation name
    /// </summary>
    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    /// <summary>
    /// Display title
    /// </summary>
    [JsonPropertyName("title")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Title { get; set; }

    /// <summary>
    /// Description of what the operation does
    /// </summary>
    [JsonPropertyName("description")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Description { get; set; }

    /// <summary>
    /// JSON Schema for input parameters
    /// </summary>
    [JsonPropertyName("input_schema")]
    public JsonElement InputSchema { get; set; }

    /// <summary>
    /// Optional output schema
    /// </summary>
    [JsonPropertyName("output_schema")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public JsonElement? OutputSchema { get; set; }

    /// <summary>
    /// Annotations (read_only, idempotent, destructive, etc.)
    /// </summary>
    [JsonPropertyName("annotations")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public JsonElement? Annotations { get; set; }

    /// <summary>
    /// Create a new operation schema
    /// </summary>
    public OperationSchema(string name, JsonElement inputSchema)
    {
        Name = name;
        InputSchema = inputSchema;
    }

    /// <summary>
    /// Set title (builder pattern)
    /// </summary>
    public OperationSchema WithTitle(string title)
    {
        Title = title;
        return this;
    }

    /// <summary>
    /// Set description (builder pattern)
    /// </summary>
    public OperationSchema WithDescription(string description)
    {
        Description = description;
        return this;
    }

    /// <summary>
    /// Set output schema (builder pattern)
    /// </summary>
    public OperationSchema WithOutputSchema(JsonElement schema)
    {
        OutputSchema = schema;
        return this;
    }

    /// <summary>
    /// Set annotations (builder pattern)
    /// </summary>
    public OperationSchema WithAnnotations(JsonElement annotations)
    {
        Annotations = annotations;
        return this;
    }
}

